File: my-folder/0399-evaluate-division/solution.py
class Solution:
    def calcEquation(self, eq: List[List[str]], val: List[float], q: List[List[str]]) -> List[float]:
    
        n = len(eq)
        graph = defaultdict(list)
        for (a,b),cost in zip(eq,val):
            graph[a].append((b,cost))
            graph[b].append((a,1/cost))

        print(graph)
        def dfs(start,end,path_cost,visited):
            
            if start==end:
                return path_cost
            visited.add(start)
            for nei,cost in graph[start]:
                if nei not in visited:
                    value = dfs(nei,end,path_cost*cost,visited)
                    if value is not None:
                        return value
            
        output = [-1]*len(q)

        for i,(start,end) in enumerate(q):
            if start==end and start in graph:
                output[i] = 1
            elif start not in graph or end not in graph:
                output[i] = -1
            else:
                out = dfs(start,end,1,set())
                if out is not None:
                    output[i] = out
        return output
            

File: my-folder/0399-evaluate-division/solution.cpp
class Solution {
public:
    unordered_map<string,vector<pair<string, double >>> graph;

    double dfs(string first,string end,double path,set<string>& visited)
    {
        if (first==end)
        {
            return path;
        }

        visited.insert(first);
        for (auto &nei: graph[first])
        {
            const string& second = nei.first;
            double weight = nei.second;
            if ( visited.find(second)== visited.end() )
            {
                double val = dfs(second,end,path*weight,visited);
                if (val!=-1)
                {
                    return val;
                }
            }
        }
        return -1.0;
    }


    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        
        int n = equations.size();
        int m = queries.size();

        for (int i =0; i<n; i++)
        {
            graph[equations[i][0]].push_back({equations[i][1],values[i]});
            graph[equations[i][1]].push_back({equations[i][0],1/values[i]});
        }

        vector<double> output(m);

        for (int i = 0; i<m; i++)
        {
            string first = queries[i][0];
            string second = queries[i][1];

            // If either node is not in the graph, the result is -1.0
            if (graph.find(first) == graph.end() || graph.find(second) == graph.end())
            {
                output[i] = -1.0;
            }
            else if (first == second)
            {
                output[i] = 1.0; // Self-equality case
            }
            else
            {
                set<string> visited;
                double out = dfs(first, second, 1.0, visited);
                output[i] = out;
            }
            
        }
        return output;


        
    }
};
